Mathlib/Logic/Encodable/Basic.lean

```diff
@@ -0,0 +1,699 @@
+-/
+
+
+open Option List Nat Function
+
+/-- Constructively countable type. Made from an explicit injection `encode : α → ℕ` and a partial
+inverse `decode : ℕ → Option α`. Note that finite types *are* countable. See `Denumerable` if you
+wish to enforce infiniteness. -/
+class Encodable (α : Type _) where
+  /-- Encoding from Type α to ℕ -/
+  encode : α → ℕ
+  --Porting note: was `decode [] : ℕ → Option α`. This means that `decode` does not take the type
+  --explicitly in Lean4
+  /-- Decoding from ℕ to Option α-/
+  decode : ℕ → Option α
+  /-- Invariant relationship between encoding and decoding-/
+  encodek : ∀ a, decode (encode a) = some a
+#align encodable Encodable
+
+attribute [simp] Encodable.encodek
+
+namespace Encodable
+
+variable {α : Type _} {β : Type _}
+
+universe u
+
+theorem encode_injective [Encodable α] : Function.Injective (@encode α _)
+  | x, y, e => Option.some.inj <| by rw [← encodek, e, encodek]
+#align encodable.encode_injective Encodable.encode_injective
+
+@[simp]
+theorem encode_inj [Encodable α] {a b : α} : encode a = encode b ↔ a = b :=
+  encode_injective.eq_iff
+#align encodable.encode_inj Encodable.encode_inj
+
+-- The priority of the instance below is less than the priorities of `Subtype.Countable`
+-- and `Quotient.Countable`
+instance (priority := 400) countable [Encodable α] : Countable α where
+  exists_injective_nat' := ⟨_,encode_injective⟩
+
+theorem surjective_decode_iget (α : Type _) [Encodable α] [Inhabited α] :
+    Surjective fun n => ((Encodable.decode n).iget : α) := fun x =>
+  ⟨Encodable.encode x, by simp_rw [Encodable.encodek]⟩
+#align encodable.surjective_decode_iget Encodable.surjective_decode_iget
+
+/-- An encodable type has decidable equality. Not set as an instance because this is usually not the
+best way to infer decidability. -/
+def decidableEqOfEncodable (α) [Encodable α] : DecidableEq α
+  | _, _ => decidable_of_iff _ encode_inj
+#align encodable.decidable_eq_of_encodable Encodable.decidableEqOfEncodable
+
+/-- If `α` is encodable and there is an injection `f : β → α`, then `β` is encodable as well. -/
+def ofLeftInjection [Encodable α] (f : β → α) (finv : α → Option β)
+    (linv : ∀ b, finv (f b) = some b) : Encodable β :=
+  ⟨fun b => encode (f b), fun n => (decode n).bind finv, fun b => by
+    simp [Encodable.encodek, linv]⟩
+#align encodable.of_left_injection Encodable.ofLeftInjection
+
+/-- If `α` is encodable and `f : β → α` is invertible, then `β` is encodable as well. -/
+def ofLeftInverse [Encodable α] (f : β → α) (finv : α → β) (linv : ∀ b, finv (f b) = b) :
+    Encodable β :=
+  ofLeftInjection f (some ∘ finv) fun b => congr_arg some (linv b)
+#align encodable.of_left_inverse Encodable.ofLeftInverse
+
+/-- Encodability is preserved by equivalence. -/
+def ofEquiv (α) [Encodable α] (e : β ≃ α) : Encodable β :=
+  ofLeftInverse e e.symm e.left_inv
+#align encodable.of_equiv Encodable.ofEquiv
+
+-- Porting note: removing @[simp], too powerful
```

Comment:
This is a `rfl`-lemma. Can `dsimp` prove this? If so, mention it in the porting note.
```suggestion
-- @[simp] -- Porting note: dsimp can prove this
```
If not please use
```suggestion
@[simp, nolint simpNF] -- Porting note: dsimp can not prove this
```

Reply:
@ChrisHughes24  -- deferring to you on this

Reply:
Or @jcommelin  -- if I replace `rfl` by `by dsimp` there's no close -- is that the test you were looking for?